### PR TITLE
fix: setting `__config_file__` and `__line__`

### PR DESCRIPTION
**Pull request by bdraco**
TLDR: Setting `__config_file__` and` __line__` on the cython object meant they were not always readable in pure python.  This turned out to be a corner case with dunder names.  

Cythonizing the `Node*Class` objects proved to be unneeded anyways since the cost of small functions is better optimized by cythonizing the constructors. Make sure they are set using the object protocol instead since there is no concern about if it will only work in Cython.

Summary of changes

- Drop cythonizing on `objects.py` since it caused the issue above
- Collapse the three different `_add_reference..` calls in `reference.py` since its all the same to `_add_reference_to_node_class` now that there is no value in cythonizing them. Add a check and test for `start_mark` being `None`
- Move `_handle_mapping_tag`, `_construct_seq`, and `_handle_scalar_tag` to `constructors.py` so they can be cythonized instead - no changes to these functions other than calls to `_add_reference_to_node_class`
- Move `_add_reference` to `reference_object.py` to avoid circular reference - changed functions calls to `_add_reference_to_node_class` and an unreachable branch that was not obvious before because the typing is incorrect here as input can be `int` as well (existing long standing problem, not something to be fixed in this PR as I didn't want to make this larger)
- Add additional test coverage

### CODE REVIEW

File: src/annotatedyaml/loader.py [hunk 1]
```diff
@@ -256,13 +252,13 @@ def _add_reference(
...
         obj = NodeListClass(obj)
-        _add_reference_to_node_list_class(obj, loader, node)
     elif isinstance(obj, str):
         obj = NodeStrClass(obj)
-        _add_reference_to_node_str_class(obj, loader, node)
     elif isinstance(obj, dict):
         obj = NodeDictClass(obj)
-        _add_reference_to_node_dict_class(obj, loader, node)
+    else:
+        return obj  # type: ignore[unreachable]
```
Comment (bdraco) [inline src/annotatedyaml/loader.py:260]: The typing on this is too narrow. It can actually be `int` as well but I didn't want to change that here since its not the substance of this PR other than the references being added.

**Review by jpbede — APPROVED** [review]
Nice! LGTM, thanks 👍

**Comment by bdraco** [comment]
thanks
